.pre-commit-config.yaml

```diff
@@ -0,0 +1,18 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+    -   id: trailing-whitespace
+    # -   id: end-of-file-fixer
+    -   id: check-yaml
+    -   id: check-added-large-files
+-   repo: https://github.com/pycqa/isort
+    rev: 5.12.0
```

Comment:
Do we really need to specify the version again, this is very prune to errors !

Have you considered not using pre-commit, but simply use .git/hooks/pre-commit.sample to run isort etc (which are already installed).

Reply:
> Do we really need to specify the version again, this is very prune to errors !

I followed the [official directions](https://pycqa.github.io/isort/docs/configuration/pre-commit.html) for `isort`, but updated the version to avoid https://github.com/PyCQA/isort/issues/2077.

We could consider using `renovatebot`, which can [supposedly](https://docs.renovatebot.com/modules/manager/pre-commit/) update `pre-commit` automatically.  (I find it much smarter than Dependabot, generally)

> Have you considered not using pre-commit, but simply use .git/hooks/pre-commit.sample to run isort etc (which are already installed).

Manually installing hooks defeats the purpose of `pre-commit`.  Ultimately it's a matter of preference, I guess.  If you don't want a shared config we can close this and add it to `.gitignore`.

Reply:
I have no problem using pre-commit, but I do have a problem with duplicated information, especially versions, that is bound to give problems. I installed the hook manually because that was the fastest for me (a simple copy) not because it is the right thing to do.

If we use pre-commit, then why have flake8 etc in requirements.txt, that makes little sense to me and it would solve the potential version conflict.

But we still need to make sure that if CI-Black is giving green light then so must pre-commit. I had for a long time the problem, that my local pylint reported a problem, and when I corrected that CI-pylint reported another problem (in the same lines), I do not want to experience that again.

Reply:
The solution here is to move CI to use `pre-commit`, and remove `flake8`/`isort`/`black`/etc. from `requirements.txt`.

That way:
- `pre-commit` solves the toolchain problem: how to (1) co-version tools, (2) standard method of kicking them all off simultaneously
- Configs still live in `setup.cfg`, as earlier
- An identical workflow used in local dev is used in CI (really beneficial)

Reply:
That is a good alternative, I like that. README.rst should still be updated to inform new developers how to setup their environment.

Reply:
> The solution here is to move CI to use `pre-commit`, and remove `flake8`/`isort`/`black`/etc. from `requirements.txt`.

Thank you, it really solved our problems here!